Add Load Balancer service functionalities.

yapoo-t opened the pull request:
This Pull Request does follwoings.

- Add Load Balancer service functionalities.
- Add unit test of the above.

yapoo-t committed:
Add Load Balancer service functionalities.

keiichi-hikita commented on ecl/network/v2/load_balancer_plans/results.go:
```diff
@@ -0,0 +1,97 @@
+	Version string `json:"version"`
+}
+
+// LoadBalancerPage is the page returned by a pager when traversing over a collection
```
Beginning of this comment block does not match to function name.

yapoo-t replied:
Thank you for reviews! I fixed.

keiichi-hikita commented on ecl/network/v2/load_balancer_plans/doc.go:
```diff
@@ -0,0 +1,25 @@
+Example to List Load Balancer Plans
+
+	listOpts := load_balancer_plans.ListOpts{
+		Status: "ACTIVE",
```
- It looks that query option of load balancer plan does not have status parameter.
- Why don't you write example of "Show Load Balancer Plans" ?(It is same in other doc.go of other resource)

yapoo-t replied:
- You're right. I fixed to the correct parameter.
- I added samples of Show method to all <resource>/doc.go

keiichi-hikita commented on ecl/network/v2/load_balancer_syslog_servers/requests.go:
```diff
@@ -0,0 +1,232 @@
+	DateFormat string `json:"date_format,omitempty"`
+
+	// Description is description
+	Description string `json:"description,omitempty"`
```
We have to consider the case that user will update the description parameter by ""(blank string).

If string parameter is defined as String&OmitEmpty, user will not be able to set blank as update parameter, because blank string will be omitted and will not be sent to API server.
For those cases, "update" parameter should be defined as "*string" if blank(=zero value in golang) is expected.

yapoo-t replied:
Thank you. I added "*" to fields "Name", "Description", "DefaultGateway" in UpdateOpts.

keiichi-hikita commented on ecl/network/v2/load_balancer_syslog_servers/doc.go:
```diff
@@ -0,0 +1,72 @@
+	loadBalancerSyslogServerID := "9ab7ab3c-38a6-417c-926b-93772c4eb2f9"
+
+	updateOpts := load_balancer_syslog_servers.UpdateOpts{
+		Name:           "new_name",
```
Load balancer syslog server does not have "name" as update parameter, right?

yapoo-t replied:
Fixed to existing parameter!

keiichi-hikita commented on ecl/network/v2/load_balancers/requests.go:
```diff
@@ -0,0 +1,183 @@
+	AdminUsername string `json:"admin_username"`
+
+	// Description is description
+	DefaultGateway string `json:"default_gateway,omitempty"`
```
Please check if user can set blank as default gateway in update case.
If it is expected case, this parameter must be defined as string pointer same as description.(Under this comment block)

kazyshr replied:
The same applies to Name field.

yapoo-t replied:
User can set blank as default gateway in update case.
I added "*" to fields "Name", "Description", "DefaultGateway" in UpdateOpts.

keiichi-hikita commented on ecl/network/v2/load_balancers/testing/doc.go:
```diff
@@ -0,0 +1,2 @@
+// ports unit tests
```
ports?

yapoo-t replied:
Fixed the doc.

keiichi-hikita commented on ecl/network/v2/load_balancer_actions/testing/doc.go:
```diff
@@ -0,0 +1,2 @@
+// ports unit tests
```
ports?

yapoo-t replied:
Fixed the doc.

keiichi-hikita commented on ecl/network/v2/load_balancer_interfaces/requests.go:
```diff
@@ -0,0 +1,151 @@
+	LoadBalancerID string `json:"load_balancer_id,omitempty"`
+
+	// Name of the Load Balancer Interface
+	Name string `json:"name,omitempty"`
```
Should be defined as string pointer same as the other resource.

yapoo-t replied:
I added "*" to fields "Name", "Description", "DefaultGateway" in UpdateOpts.

keiichi-hikita commented on ecl/network/v2/load_balancer_interfaces/results.go:
```diff
@@ -0,0 +1,112 @@
+	VirtualIPAddress string `json:"virtual_ip_address"`
+
+	// Properties used for virtual IP address
+	VirtualIPProperties VirtualIpProperties `json:"virtual_ip_properties"`
```
You do not need to define VirtualIpProperties to avoid name duplication.
I think you can define VirtualIPProperties struct and can use here as variable type of "VirtualIPProperties" also.

yapoo-t replied:
Fixed. 
Sorry for that, I was supposed to rename to VirtualIPProperties, but there was omission of correction.

keiichi-hikita commented on ecl/network/v2/load_balancer_interfaces/testing/doc.go:
```diff
@@ -0,0 +1,2 @@
+// ports unit tests
```
ports

yapoo-t replied:
Fixed!

kazyshr commented on ecl/network/v2/load_balancers/requests.go:
```diff
@@ -0,0 +1,183 @@
+type UpdateOpts struct {
+
+	// AdminUsername is admin's username
+	AdminUsername string `json:"admin_username"`
```
This parameter cannot be updated.
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer.html#id31

yapoo-t replied:
Thank you for reviews. I removed the field.

kazyshr commented on ecl/network/v2/load_balancers/requests.go:
```diff
@@ -0,0 +1,183 @@
+		r.Err = err
+		return
+	}
+	_, r.Err = c.Post(createURL(c), b, &r.Body, nil)
```
You need to set OKCodes because this endpoint returns 200 as success code. It is different from default codes of POST method. Please look at `defaultOkCodes` function.
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer.html#id9
```suggestion
	_, r.Err = c.Post(createURL(c), b, &r.Body, &eclcloud.RequestOpts{
		OkCodes: []int{200},
	})
```

yapoo-t replied:
Thank you.
There are same problems in below so I will fix them:
  load_balancer_syslog_servers.Create
  load_balancer_actions.Reboot
  load_balancer_actions.ResetPassword

kazyshr commented on ecl/network/v2/load_balancers/requests.go:
```diff
@@ -0,0 +1,183 @@
+		return
+	}
+	_, r.Err = c.Put(updateURL(c, id), b, &r.Body, &eclcloud.RequestOpts{
+		OkCodes: []int{200, 201},
```
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer.html#id33
```suggestion
		OkCodes: []int{200},
```

yapoo-t replied:
Thank you.
There are same problems in below so I will fix them:
load_balancer_syslog_servers.Update
load_balancer_interfaces.Update

kazyshr commented on ecl/network/v2/load_balancers/requests.go:
```diff
@@ -0,0 +1,183 @@
+	DefaultGateway     string `q:"default_gateway"`
+	Description        string `q:"description"`
+	ID                 string `q:"id"`
+	LoadBalancerPlanId string `q:"load_balancer_plan_id"`
```
This filed should be changed for consistency.
```suggestion
	LoadBalancerPlanID string `q:"load_balancer_plan_id"`
```

yapoo-t replied:
Thank you for suggestion!

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,151 @@
+// NextPageURL is invoked when a paginated collection of load balancers has reached
+// the end of a page and the pager seeks to traverse over a new one. In order
+// to do this, it needs to construct the next page's URL.
+func (r LoadBalancerPage) NextPageURL() (string, error) {
```
The API of load balancer does not support pagination. So this method should be removed.

yapoo-t replied:
Got it, I fixed. I also fixed below files:
ecl/network/v2/load_balancer_interfaces/results.go
ecl/network/v2/load_balancer_plans/results.go
ecl/network/v2/load_balancer_syslog_servers/results.go

yapoo-t committed:
Add some changes requested by keiichi-hikita and kazyshr (review comments until Jan.29 17:37 JST)

- Add samples of Show method to <resource>/doc.go
- Fix sample codes of <resource>/doc.go
- Fix description of <resource>/testing/doc.go
- Add "*" to fields "Name", "Description", "DefaultGateway" in UpdateOpts
- Remove unnecessary fields in UpdateOpts of load_balancer_interfaces package.
- Rename structure name "VirtualIpProperties" to "VirtualIPProperties"
- Fix comments of method "LoadBalancerPlanPage" in load_balancer_plans/results.go
- Remove unnecessary field "AdminUsername" in UpdateOpts of load_balancers package.

yapoo-t committed:
Update ecl/network/v2/load_balancers/requests.go

Set OKCodes

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancers/requests.go

Fix OkCodes in Update method.

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancers/requests.go

Fix field name

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Fix status codes in Create and Update method

load_balancer_syslog_servers.Create
load_balancer_actions.Reboot
load_balancer_actions.ResetPassword
load_balancer_syslog_servers.Update
load_balancer_interfaces.Update

yapoo-t committed:
Remove NextPageURL methods.

keiichi-hikita reviewed (approved):
Thanks for your response!
LGTM o/

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,137 @@
+	AvailabilityZone string `json:"availability_zone"`
+
+	// Description is description
+	DefaultGateway string `json:"default_gateway"`
```
Default value of default_gateway is null. So you need to distinguish between nil and blank string.
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer.html#id6
```suggestion
	DefaultGateway *string `json:"default_gateway"`
```

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,137 @@
+	SlotNumber          int    `json:"slot_number"`
+	Status              string `json:"status"`
+	Type                string `json:"type"`
+	VirtualIPAddress    string `json:"virtual_ip_address"`
```
Seems like default value is null though I cannot find official documents.
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer.html#id14
```suggestion
	VirtualIPAddress    *string `json:"virtual_ip_address"`
```

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,137 @@
+	Status              string `json:"status"`
+	Type                string `json:"type"`
+	VirtualIPAddress    string `json:"virtual_ip_address"`
+	VirtualIPProperties string `json:"virtual_ip_properties"`
```
```suggestion
	VirtualIPProperties *string `json:"virtual_ip_properties"`
```

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,137 @@
+	ID                  string `json:"id"`
+	IPAddress           string `json:"ip_address"`
+	Name                string `json:"name"`
+	NetworkID           string `json:"network_id"`
```
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer.html#id26
```suggestion
	NetworkID           *string `json:"network_id"`
```

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,137 @@
+// Interface represents a attached interfaces by Load Balancer.
+type Interface struct {
+	ID                  string `json:"id"`
+	IPAddress           string `json:"ip_address"`
```
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer.html#id26
```suggestion
	IPAddress           *string `json:"ip_address"`
```

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+// Interface represents a attached interfaces by Load Balancer.
+type Interface struct {
```
You can use LoadBalancerInterface type in load_balancer_interfaces package instead of defining this here.

yapoo-t replied:
Thanks. Fixed the type of field to LoadBalancerInterface.

kazyshr commented on ecl/network/v2/load_balancers/results.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+// Interface represents a attached interfaces by Load Balancer.
+type SyslogServer struct {
```
You can use LoadBalancerSyslogServer type in load_balancer_syslog_servers package instead of defining this here.

yapoo-t replied:
Thanks. Fixed the type of field to LoadBalancerSyslogServer .

kazyshr commented on ecl/network/v2/load_balancers/requests.go:
```diff
@@ -0,0 +1,182 @@
+
+// IDFromName is a convenience function that returns a Load Balancer's ID,
+// given its name.
+func IDFromName(client *eclcloud.ServiceClient, name string) (string, error) {
```
Can you write unit tests for this function or remove it if it is not used in terraform? Please check the function in the other packages.

yapoo-t replied:
OK, I wrote unit tests for the function IDFromName.

kazyshr commented on ecl/network/v2/load_balancer_syslog_servers/requests.go:
```diff
@@ -0,0 +1,234 @@
+	PortNumber int `json:"port_number,omitempty"`
+
+	// priority (0-255)
+	Priority int `json:"priority,omitempty"`
```
If user sets 0 in Priority field, it will be omitted. So this should be pointer.
```suggestion
	Priority *int `json:"priority,omitempty"`
```

kazyshr commented on ecl/network/v2/load_balancer_syslog_servers/requests.go:
```diff
@@ -0,0 +1,234 @@
+	LogLevel string `json:"log_level,omitempty"`
+
+	// priority (0-255)
+	Priority int `json:"priority,omitempty"`
```
```suggestion
	Priority *int `json:"priority,omitempty"`
```

kazyshr replied:
It was not fixed yet.

kazyshr commented on ecl/network/v2/load_balancer_interfaces/results.go:
```diff
@@ -0,0 +1,98 @@
+	ID string `json:"id"`
+
+	// IP Address
+	IPAddress string `json:"ip_address"`
```
Seems like this filed can have null value.
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer_interface.html#sample-response-body
```suggestion
	IPAddress *string `json:"ip_address"`
```

kazyshr commented on ecl/network/v2/load_balancer_interfaces/results.go:
```diff
@@ -0,0 +1,98 @@
+	Name string `json:"name"`
+
+	// UUID of the parent network.
+	NetworkID string `json:"network_id"`
```
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer_interface.html#sample-response-body
```suggestion
	NetworkID *string `json:"network_id"`
```

kazyshr commented on ecl/network/v2/load_balancer_interfaces/results.go:
```diff
@@ -0,0 +1,98 @@
+	TenantID string `json:"tenant_id"`
+
+	// Virtual IP Address
+	VirtualIPAddress string `json:"virtual_ip_address"`
```
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer_interface.html#sample-response-body
```suggestion
	VirtualIPAddress *string `json:"virtual_ip_address"`
```

kazyshr commented on ecl/network/v2/load_balancer_interfaces/results.go:
```diff
@@ -0,0 +1,98 @@
+	VirtualIPAddress string `json:"virtual_ip_address"`
+
+	// Properties used for virtual IP address
+	VirtualIPProperties VirtualIPProperties `json:"virtual_ip_properties"`
```
https://ecl.ntt.com/en/documents/api-references/network/load_balancer/load_balancer_interface.html#sample-response-body
```suggestion
	VirtualIPProperties *VirtualIPProperties `json:"virtual_ip_properties"`
```

kazyshr commented on ecl/network/v2/load_balancer_actions/results.go:
```diff
@@ -0,0 +1,38 @@
+}
+
+// LoadBalancerActionResetPassword represents a detail of a Reset Password operation.
+type LoadBalancerActionResetPassword struct {
```
I prefer shorter name like this.
```suggestion
type ResetPassword struct {
```

yapoo-t replied:
`ResetPassword` : conflicted to the method name using this structure.
  -> changed to `ResetPasswordDetail`

kazyshr replied:
Sorry, I did not realize name conflict. So how is `Password` simply?

yapoo-t replied:
Nice name! Changed as you suggested.

kazyshr reviewed (changes requested):
Great job!! I added some comments. 
Can you run `make fmt` to format codes?

yapoo-t committed:
Fix structure field type in load_balancers package

- Use LoadBalancerInterface type in load_balancer_interfaces package instead of defining in load_balancers package
- Use LoadBalancerSyslogServer  type in load_balancer_syslog_servers package instead of defining in load_balancers package

yapoo-t committed:
Update ecl/network/v2/load_balancer_syslog_servers/requests.go

Change field type to pointer

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancer_syslog_servers/requests.go

Change field type to pointer

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancer_interfaces/results.go

Change field type to pointer

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancer_interfaces/results.go

Change field type to pointer

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancer_interfaces/results.go

Change field type to pointer

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancer_interfaces/results.go

Change field type to pointer

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancer_actions/results.go

Shorten type name.

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Update ecl/network/v2/load_balancers/results.go

Change field type to pointer

Co-Authored-By: kazyshr <30496953+kazyshr@users.noreply.github.com>

yapoo-t committed:
Fixed tests related to modification of field types to pointer

yapoo-t committed:
Fixed doc related to modification of field types to pointer

yapoo-t committed:
Fixed shortened type name conflicted with other method name.

- Also fixed test codes using the type.

yapoo-t committed:
Run make fmt

yapoo-t committed:
Fixed error message in test codes.

yapoo-t committed:
Write unit tests for func IDFromName

yapoo-t commented:
> Great job!! I added some comments.

Thank you for your detailed reviews!
I saw all your comments and applied changes.

> Can you run `make fmt` to format codes?

Done!

yapoo-t committed:
Changed structure name ResetPasswordDetail to Password

kazyshr reviewed (approved):
I confirmed that all comments got fixed. Thanks!!
